File: combine_sboms.py
import json
import sys
import os
import uuid
from datetime import datetime
from typing import List, Dict, Set, Tuple

def read_sbom(sbom_file: str, debug: bool = False) -> Tuple[Dict, str]:
    try:
        encodings = ['utf-8', 'utf-8-sig', 'latin1']
        sbom_data = None
        for encoding in encodings:
            try:
                with open(sbom_file, 'r', encoding=encoding) as file:
                    sbom_data = json.load(file)
                if debug:
                    print(f"Successfully read {sbom_file} with encoding: {encoding}")
                break
            except UnicodeDecodeError:
                continue
        if sbom_data is None:
            print(f"Error: Unable to decode '{sbom_file}' with supported encodings (utf-8, utf-8-sig, latin1)")
            return {}, "unknown"
        
        if "bomFormat" in sbom_data and sbom_data["bomFormat"] == "CycloneDX":
            sbom_format = "CycloneDX"
        elif "spdxVersion" in sbom_data:
            sbom_format = "SPDX"
        else:
            print(f"Unsupported SBOM format in {sbom_file}. Expected CycloneDX or SPDX.")
            return {}, "unknown"
        
        return sbom_data, sbom_format
    
    except FileNotFoundError:
        print(f"Error: File '{sbom_file}' not found")
        return {}, "unknown"
    except json.JSONDecodeError as e:
        print(f"Error: '{sbom_file}' is not a valid JSON file: {str(e)}")
        return {}, "unknown"
    except Exception as e:
        print(f"Error processing SBOM file {sbom_file}: {str(e)}")
        return {}, "unknown"

def get_metadata_from_sbom(sbom_data: Dict, sbom_format: str, debug: bool = False) -> Tuple[str, str]:
    name = "Combined Project"
    version = "1.0.0"
    
    if sbom_format == "CycloneDX":
        if "metadata" in sbom_data and "component" in sbom_data["metadata"]:
            name = sbom_data["metadata"]["component"].get("name", name)
            version = sbom_data["metadata"]["component"].get("version", version)
            if debug:
                print(f"Extracted metadata from CycloneDX: name={name}, version={version}")
    elif sbom_format == "SPDX":
        name = sbom_data.get("name", name)
        if "metadata" in sbom_data:
            name = sbom_data["metadata"].get("name", name)
            version = sbom_data["metadata"].get("versionInfo", version)
        if debug:
            print(f"Extracted metadata from SPDX: name={name}, version={version}")
    
    return name, version

def convert_spdx_to_cyclonedx_package(package: Dict, debug: bool = False) -> Dict:
    component = {
        "type": "library",  # Default type, as SPDX doesn't always specify
        "name": package.get("name", "unnamed"),
        "version": package.get("versionInfo", "unknown"),
        "bom-ref": package.get("SPDXID", str(uuid.uuid4())),
    }
    licenses = []
    license_concluded = package.get("licenseConcluded")
    license_declared = package.get("licenseDeclared")
    if license_concluded and license_concluded != "NOASSERTION":
        licenses.append({"license": {"id": license_concluded}})
        if debug:
            print(f"Converted SPDX licenseConcluded {license_concluded} to CycloneDX format for {component['name']}")
    if license_declared and license_declared != "NOASSERTION":
        licenses.append({"license": {"id": license_declared}})
        if debug:
            print(f"Converted SPDX licenseDeclared {license_declared} to CycloneDX format for {component['name']}")
    if licenses:
        component["licenses"] = licenses
    return component

def combine_sboms(sbom_files: List[str], project_name: str, project_version: str, debug: bool = False) -> Tuple[Dict, str, str]:
    combined_components: Dict[Tuple[str, str], Dict] = {}
    metadata_name = project_name
    metadata_version = project_version
    
    # Extract metadata from the first valid SBOM if not provided
    if not project_name or not project_version:
        for sbom_file in sbom_files:
            sbom_data, sbom_format = read_sbom(sbom_file, debug)
            if sbom_data:
                name, version = get_metadata_from_sbom(sbom_data, sbom_format, debug)
                if not project_name:
                    metadata_name = name
                if not project_version:
                    metadata_version = version
                break
        else:
            print("Warning: No valid SBOMs found to extract metadata. Using defaults.")
    
    # Combine components
    for sbom_file in sbom_files:
        sbom_data, sbom_format = read_sbom(sbom_file, debug)
        if not sbom_data:
            continue
        
        if sbom_format == "CycloneDX":
            components = sbom_data.get("components", [])
            for component in components:
                name = component.get("name", "unnamed")
                version = component.get("version", "unknown")
                key = (name, version)
                if key not in combined_components:
                    combined_components[key] = component
                    if debug:
                        print(f"Added CycloneDX component: {name} {version} from {sbom_file}")
                elif debug:
                    print(f"Skipped duplicate component: {name} {version} from {sbom_file}")
        
        elif sbom_format == "SPDX":
            packages = sbom_data.get("packages", [])
            for package in packages:
                component = convert_spdx_to_cyclonedx_package(package, debug)
                name = component["name"]
                version = component["version"]
                key = (name, version)
                if key not in combined_components:
                    combined_components[key] = component
                    if debug:
                        print(f"Added converted SPDX package: {name} {version} from {sbom_file}")
                elif debug:
                    print(f"Skipped duplicate package: {name} {version} from {sbom_file}")
    
    # Create combined CycloneDX SBOM
    combined_sbom = {
        "bomFormat": "CycloneDX",
        "specVersion": "1.4",
        "version": 1,
        "serialNumber": f"urn:uuid:{str(uuid.uuid4())}",
        "metadata": {
            "timestamp": datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ"),
            "component": {
                "type": "application",
                "name": metadata_name,
                "version": metadata_version
            }
        },
        "components": list(combined_components.values())
    }
    
    if debug:
        print(f"Combined {len(combined_components)} unique components into SBOM")
    
    return combined_sbom, metadata_name, metadata_version

def main():
    debug = False
    project_name = ""
    project_version = ""
    
    args = sys.argv[1:]
    if "--debug" in args:
        debug = True
        args.remove("--debug")
    
    if "--name" in args:
        name_index = args.index("--name")
        if name_index + 1 < len(args):
            project_name = args[name_index + 1]
            args = args[:name_index] + args[name_index + 2:]
        else:
            print("Error: --name requires a project name")
            sys.exit(1)
    
    if "--version" in args:
        version_index = args.index("--version")
        if version_index + 1 < len(args):
            project_version = args[version_index + 1]
            args = args[:version_index] + args[version_index + 2:]
        else:
            print("Error: --version requires a version")
            sys.exit(1)
    
    if not args:
        print("Usage: python combine_sboms.py [--debug] [--name <project_name>] [--version <version>] <sbom_file1> <sbom_file2> ...")
        sys.exit(1)
    
    sbom_files = args
    for sbom_file in sbom_files:
        if not os.path.isfile(sbom_file):
            print(f"Error: File '{sbom_file}' does not exist")
            sys.exit(1)
    
    # Combine SBOMs
    combined_sbom, metadata_name, metadata_version = combine_sboms(sbom_files, project_name, project_version, debug)
    
    # Save combined SBOM
    output_file = "combined_sbom.json"
    try:
        with open(output_file, 'w', encoding='utf-8') as file:
            json.dump(combined_sbom, file, indent=2)
        print(f"Combined SBOM saved to {output_file} (name: {metadata_name}, version: {metadata_version})")
    except Exception as e:
        print(f"Error writing to {output_file}: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: src/combine_sboms.ts
#!/usr/bin/env node
import { randomUUID } from 'crypto'
import { existsSync } from 'fs'
import { readFile, writeFile } from 'fs/promises'

interface Component {
    type: string
    name: string
    group?: string
    version: string
    'bom-ref'?: string
    author?: string
    description?: string
    licenses?: Array<{ license: { id: string } } | { expression: string }>
    purl?: string
    externalReferences?: Array<{ url: string; type: string; comment: string }>
    properties?: Array<{ name: string; value: string }>
}

interface SbomData {
    bomFormat?: string
    spdxVersion?: string
    components?: Component[]
    packages?: Array<{
        name: string
        versionInfo: string
        SPDXID?: string
        licenseConcluded?: string
        licenseDeclared?: string
    }>
    metadata?: {
        component?: { name?: string; version?: string }
        name?: string
        versionInfo?: string
    }
    name?: string
}

interface CombinedSbom {
    bomFormat: string
    specVersion: string
    version: number
    serialNumber: string
    metadata: {
        timestamp: string
        component: { type: string; name: string; version: string }
    }
    components: Component[]
}

function sanitizeComponent(component: Component): Component {
    return {
        type: component.type,
        name: component.name,
        group: component.group,
        version: component.version,
        'bom-ref': component['bom-ref'],
        author: component.author,
        description: component.description,
        purl: component['purl'],
        externalReferences: component.externalReferences,
        properties: component.properties,
        licenses: component.licenses
            ?.map((lic, idx) => {
                if ('license' in lic && lic.license?.id) {
                    return {
                        license: {
                            id: lic.license.id,
                        },
                    }
                } else if ('expression' in lic && typeof lic.expression === 'string') {
                    return {
                        expression: lic.expression,
                    }
                } else {
                    console.warn(
                        `⚠️ Skipping malformed license in component ${component.name}@${component.version} (index ${idx})`
                    )
                    return undefined
                }
            })
            .filter((l): l is { license: { id: string } } | { expression: string } => !!l),
    }
}

async function readSbom(sbomFile: string, debug: boolean = false): Promise<[SbomData, string]> {
    try {
        // Try UTF-8 first
        let fileContent: string
        try {
            fileContent = await readFile(sbomFile, { encoding: 'utf8' })
            if (debug) {
                console.log(`Successfully read ${sbomFile} with encoding: utf-8`)
            }
        } catch (error: any) {
            if (error.code === 'ERR_INVALID_CHAR') {
                // Fallback to UTF-8-BOM or similar by stripping BOM
                const buffer = await readFile(sbomFile)
                fileContent = buffer.toString('utf8').replace(/^\uFEFF/, '')
                if (debug) {
                    console.log(`Successfully read ${sbomFile} with BOM-stripped utf-8`)
                }
            } else {
                throw error
            }
        }

        const sbomData: SbomData = JSON.parse(fileContent)

        if (sbomData.bomFormat === 'CycloneDX') {
            return [sbomData, 'CycloneDX']
        } else if (sbomData.spdxVersion) {
            return [sbomData, 'SPDX']
        } else {
            console.log(`Unsupported SBOM format in ${sbomFile}. Expected CycloneDX or SPDX.`)
            return [{}, 'unknown']
        }
    } catch (error: any) {
        if (error.code === 'ENOENT') {
            console.log(`Error: File '${sbomFile}' not found`)
        } else if (error instanceof SyntaxError) {
            console.log(`Error: '${sbomFile}' is not a valid JSON file: ${error.message}`)
        } else {
            console.log(`Error processing SBOM file ${sbomFile}: ${error.message}`)
        }
        return [{}, 'unknown']
    }
}

function getMetadataFromSbom(sbomData: SbomData, sbomFormat: string, debug: boolean = false): [string, string] {
    let name = 'Combined Project'
    let version = '1.0.0'

    if (sbomFormat === 'CycloneDX') {
        if (sbomData.metadata?.component) {
            name = sbomData.metadata.component.name ?? name
            version = sbomData.metadata.component.version ?? version
            if (debug) {
                console.log(`Extracted metadata from CycloneDX: name=${name}, version=${version}`)
            }
        }
    } else if (sbomFormat === 'SPDX') {
        name = sbomData.name ?? name
        if (sbomData.metadata) {
            name = sbomData.metadata.name ?? name
            version = sbomData.metadata.versionInfo ?? version
        }
        if (debug) {
            console.log(`Extracted metadata from SPDX: name=${name}, version=${version}`)
        }
    }

    return [name, version]
}

function convertSpdxToCycloneDxPackage(
    packageData: NonNullable<SbomData['packages']>[number],
    debug: boolean = false
): Component {
    const component: Component = {
        type: 'library',
        name: packageData.name ?? 'unnamed',
        version: packageData.versionInfo ?? 'unknown',
        'bom-ref': packageData.SPDXID ?? randomUUID(),
    }

    const licenses: Array<{ license: { id: string } }> = []
    if (packageData.licenseConcluded && packageData.licenseConcluded !== 'NOASSERTION') {
        licenses.push({ license: { id: packageData.licenseConcluded } })
        if (debug) {
            console.log(
                `Converted SPDX licenseConcluded ${packageData.licenseConcluded} to CycloneDX format for ${component.name}`
            )
        }
    }
    if (packageData.licenseDeclared && packageData.licenseDeclared !== 'NOASSERTION') {
        licenses.push({ license: { id: packageData.licenseDeclared } })
        if (debug) {
            console.log(
                `Converted SPDX licenseDeclared ${packageData.licenseDeclared} to CycloneDX format for ${component.name}`
            )
        }
    }
    if (licenses.length > 0) {
        component.licenses = licenses
    }

    return component
}

async function combineSboms(
    sbomFiles: string[],
    projectName: string,
    projectVersion: string,
    debug: boolean = false,
    trace: boolean = false
): Promise<[CombinedSbom, string, string]> {
    const combinedComponents: { [key: string]: Component } = {}
    let metadataName = projectName
    let metadataVersion = projectVersion

    // Extract metadata from the first valid SBOM if not provided
    if (!projectName || !projectVersion) {
        for (const sbomFile of sbomFiles) {
            const [sbomData, sbomFormat] = await readSbom(sbomFile, debug)
            if (Object.keys(sbomData).length > 0) {
                const [name, version] = getMetadataFromSbom(sbomData, sbomFormat, debug)
                if (!projectName) metadataName = name
                if (!projectVersion) metadataVersion = version
                break
            }
        }
        if (!metadataName || !metadataVersion) {
            console.log('Warning: No valid SBOMs found to extract metadata. Using defaults.')
        }
    }

    for (const sbomFile of sbomFiles) {
        const [sbomData, sbomFormat] = await readSbom(sbomFile, debug)
        if (debug) {
            console.log(
                `📄 ${sbomFile} contains components:`,
                Array.isArray(sbomData.components),
                sbomData.components?.length ?? 0
            )
        }
        if (Object.keys(sbomData).length === 0) {
            if (debug) {
                console.log(`📦 ${sbomFile} has ${sbomData.components?.length ?? 0} components`)
            }
            continue
        }

        if (sbomFormat === 'CycloneDX') {
            const components = sbomData.components ?? []
            for (const component of components) {
                const name = component.name ?? 'unnamed'
                const version = component.version ?? 'unknown'
                const key = `${name}:${version}`

                if (component.licenses) {
                    for (const lic of component.licenses) {
                        if ('acknowlegement' in (lic ?? {})) {
                            console.log(
                                `⚠️ Found unexpected 'acknowlegement' in component ${name}@${version} from ${sbomFile}`
                            )
                        }
                        if (!lic) {
                            console.warn(`⚠️ License object missing in component ${name}@${version} from ${sbomFile}`)
                        }
                    }
                }

                if (!combinedComponents[key]) {
                    combinedComponents[key] = sanitizeComponent(component)
                    if (trace) {
                        console.log(`Added CycloneDX component: ${name} ${version} from ${sbomFile}`)
                    }
                } else if (trace) {
                    console.log(`Skipped duplicate component: ${name} ${version} from ${sbomFile}`)
                }
            }
        } else if (sbomFormat === 'SPDX') {
            const packages = sbomData.packages ?? []
            for (const pkg of packages) {
                const component = convertSpdxToCycloneDxPackage(pkg, debug)
                const name = component.name
                const version = component.version
                const key = `${name}:${version}`
                if (!combinedComponents[key]) {
                    combinedComponents[key] = sanitizeComponent(component)
                    if (trace) {
                        console.log(`Added converted SPDX package: ${name} ${version} from ${sbomFile}`)
                    }
                } else if (trace) {
                    console.log(`Skipped duplicate package: ${name} ${version} from ${sbomFile}`)
                }
            }
        }
    }

    const combinedSbom: CombinedSbom = {
        bomFormat: 'CycloneDX',
        specVersion: '1.4',
        version: 1,
        serialNumber: `urn:uuid:${randomUUID()}`,
        metadata: {
            timestamp: new Date().toISOString(),
            component: {
                type: 'application',
                name: metadataName,
                version: metadataVersion,
            },
        },
        components: Object.values(combinedComponents),
    }

    if (debug) {
        console.log(`Combined ${Object.keys(combinedComponents).length} unique components into SBOM`)
    }

    return [combinedSbom, metadataName, metadataVersion]
}

async function main(): Promise<void> {
    let debug = false
    let trace = false
    let projectName = ''
    let projectVersion = ''

    const args = process.argv.slice(2)
    if (args.includes('--debug')) {
        debug = true
        args.splice(args.indexOf('--debug'), 1)
    }

    if (args.includes('--trace')) {
        debug = true
        trace = true
        args.splice(args.indexOf('--trace'), 1)
    }

    let nameIndex = args.indexOf('--name')
    if (nameIndex !== -1) {
        if (nameIndex + 1 < args.length) {
            projectName = args[nameIndex + 1]
            args.splice(nameIndex, 2)
        } else {
            console.log('Error: --name requires a project name')
            process.exit(1)
        }
    }

    let versionIndex = args.indexOf('--version')
    if (versionIndex !== -1) {
        if (versionIndex + 1 < args.length) {
            projectVersion = args[versionIndex + 1]
            args.splice(versionIndex, 2)
        } else {
            console.log('Error: --version requires a version')
            process.exit(1)
        }
    }

    if (args.length === 0) {
        console.log(
            'Usage: tsx combine_sboms.ts [--debug] [--trace] [--name <project_name>] [--version <version>] <sbom_file1> <sbom_file2> ...'
        )
        process.exit(1)
    }

    const sbomFiles = args
    for (const sbomFile of sbomFiles) {
        if (!existsSync(sbomFile)) {
            console.log(`Error: File '${sbomFile}' does not exist`)
            process.exit(1)
        }
    }

    const [combinedSbom, metadataName, metadataVersion] = await combineSboms(
        sbomFiles,
        projectName,
        projectVersion,
        debug,
        trace
    )

    const outputFile = 'combined_sbom.json'
    try {
        await writeFile(outputFile, JSON.stringify(combinedSbom, null, 2), {
            encoding: 'utf-8',
        })
        console.log(`Combined SBOM saved to ${outputFile} (name: ${metadataName}, version: ${metadataVersion})`)
    } catch (error: any) {
        console.log(`Error writing to ${outputFile}: ${error.message}`)
        process.exit(1)
    }
}

main().catch((error) => {
    console.error(`Error: ${error.message}`)
    process.exit(1)
})
